### PR TITLE
Implement Use Case Selection screen

### PR DESCRIPTION
**Pull request by justjanne**
Type: Enhancement
Related: https://github.com/vector-im/element-web/issues/22463

---
Here's what your changelog entry will look like:

## ✨ Features
 * Implement Use Case Selection screen ([\#8984](https://github.com/matrix-org/matrix-react-sdk/pull/8984)). Contributed by @justjanne.

### CODE REVIEW

**Comment by justjanne** [comment]
TODO: The current positioning of it being post-login is just for testing, before I move it out of draft it’ll be moved, as intended, to a post-registration action

**Comment by justjanne** [comment]
I’ve extracted all the relevant parts from that video, to show which parts show in which themes:

* Previous page, "verify your email" screen, light mode on mountain background  
  
* Browser waiting for page to load, dark background   
  
* Browser has loaded an empty page, white background  
  frame-41
* Default content for auth route, login screen in white on white background (neither dark mode css nor mountain background have loaded yet)
  
* Default content for auth route, login screen in dark mode on black background (mountain background hasn’t loaded yet) 
  
* Setting up keys screen, dark mode on black background (mountain background hasn’t loaded yet) 
  
* Setting up keys screen with setting up keys modal, dark mode on mountain background
  
* Setting up keys screen, dark mode on mountain background
  
* New use case selection screen (no content showing), dark mode on dark gradient background
  
* New use case selection screen (only title showing), dark mode on dark gradient background
  
* New use case selection screen (all content showing), dark mode on dark gradient background

File: res/css/structures/_SplashPage.scss
Comment (turt2live) [inline res/css/structures/_SplashPage.scss:1]: file needs copyright

File: res/css/views/dialogs/_UseCaseSelection.scss
Comment (turt2live) [inline res/css/views/dialogs/_UseCaseSelection.scss:1]: file needs copyright

File: res/css/views/dialogs/_UseCaseSelection.scss
Comment (turt2live) [inline res/css/views/dialogs/_UseCaseSelection.scss:13]: needs to be resolved before merge

File: res/css/views/dialogs/_UseCaseSelection.scss
Comment (turt2live) [inline res/css/views/dialogs/_UseCaseSelection.scss:28]: needs to be resolved before merge

File: res/css/views/dialogs/_UseCaseSelection.scss
Comment (turt2live) [inline res/css/views/dialogs/_UseCaseSelection.scss:37]: needs to be resolved before merge

File: res/css/views/dialogs/_UseCaseSelection.scss
Comment (turt2live) [inline res/css/views/dialogs/_UseCaseSelection.scss:49]: needs to be resolved before merge

File: res/css/views/dialogs/_UseCaseSelection.scss
Comment (turt2live) [inline res/css/views/dialogs/_UseCaseSelection.scss:66]: needs to be resolved before merge

File: res/css/views/dialogs/_UseCaseSelection.scss
Comment (turt2live) [inline res/css/views/dialogs/_UseCaseSelection.scss:73]: needs to be resolved before merge

File: src/components/structures/MatrixChat.tsx
Comment (turt2live) [inline src/components/structures/MatrixChat.tsx:1215]: I don't think the setting level check is needed here, and the formatting makes the condition bleed into the content:

```suggestion
        if (
            MatrixClientPeg.currentUserIsJustRegistered() &&
            SettingsStore.getValue("ftueUseCaseSelection") === null
        ) {
```

File: src/components/structures/MatrixChat.tsx
Comment (turt2live) [inline src/components/structures/MatrixChat.tsx:1221]: we should also clean up the watcher here, but instead can we use the dispatcher? The settings store isn't really meant for this ephemeral signaling type stuff.

File: src/components/views/dialogs/UseCaseSelection.tsx
Comment (turt2live) [inline src/components/views/dialogs/UseCaseSelection.tsx:4]: this copyright seems incorrect - should only be 2022 Foundation

File: src/components/views/dialogs/UseCaseSelection.tsx
Comment (turt2live) [inline src/components/views/dialogs/UseCaseSelection.tsx:40]: applies throughout: we replace the fancy quotes (`’`) with un-fancy ones (`'`) for reasons relating to translations and consistency, though if design says they should stay then they stay.

File: src/settings/Settings.tsx
Comment (turt2live) [inline src/settings/Settings.tsx:701]: For naming guidelines on settings:
```suggestion
    "FTUE.useCaseSelection": {
```

will need to be applied elsewhere

File: src/components/views/dialogs/UseCaseSelection.tsx
Comment (turt2live) [inline src/components/views/dialogs/UseCaseSelection.tsx:30]: this enum should be its own file since it's exported from a component

File: src/components/views/dialogs/UseCaseSelection.tsx
Comment (turt2live) [inline src/components/views/dialogs/UseCaseSelection.tsx:29]:
```suggestion
    Skip = "Skip",
```

File: src/settings/Settings.tsx
Comment (turt2live) [inline src/settings/Settings.tsx:702]: so it can be overridden safely:
```suggestion
        supportedLevels: LEVELS_ACCOUNT_SETTINGS,
```

Comment (turt2live) [reply src/settings/Settings.tsx:702]: this will also be the way for managed deployments to disable it.

**Review by turt2live — CHANGES_REQUESTED** [review]
Generally this lgtm, but it doesn't seem to be informing anything yet? Should this be behind a labs flag while we build out the stuff that uses the selection for something?

File: src/components/structures/MatrixChat.tsx
Comment (justjanne) [reply src/components/structures/MatrixChat.tsx:1223]: I’ll refer to the comments on the posthog analytics toast (which is happening after this code runs), which explains why one should use a watcher here:

https://github.com/matrix-org/matrix-react-sdk/blob/8bb1c7567f69f8c5b7a79876a3d6848af61d17bc/src/components/structures/MatrixChat.tsx#L1290-L1305

**Comment by justjanne** [comment]
> Generally this lgtm, but it doesn't seem to be informing anything yet? Should this be behind a labs flag while we build out the stuff that uses the selection for something?

Our mobile apps have already implemented this screen. The primary benefit of this for now is that we immediately get the posthog metrics of which group a user belongs to.

Actual user-facing changes will happen with the new user onboarding checklist later on.

**Comment by niquewoodhouse** [comment]
> This is already the minimal amount of banding possible unless we add random noise to the gradient.
> 
> I don’t think there’s anything we could do about the UI jumping around in this part of the code, we’d probably have to modify the other parts of the UI for this. The only thing I could try to avoid would be the key UI.

Just to ask, is modifying other parts ok to do?  I'm concerned this amount of jumping is going to create a negative idea of element in new user's minds (woah...this seems broken/janky).  Alternatively, we could match the use case more to other UI but would rather not show it at all.

**Comment by turt2live** [comment]
> > Generally this lgtm, but it doesn't seem to be informing anything yet? Should this be behind a labs flag while we build out the stuff that uses the selection for something?
> 
> Our mobile apps have already implemented this screen. The primary benefit of this for now is that we immediately get the posthog metrics of which group a user belongs to.
> 
> Actual user-facing changes will happen with the new user onboarding checklist later on.

has Product signed off on this? I'm a bit concerned it affects the user experience in a net-negative way to not have follow up questions at the moment.

File: src/components/structures/MatrixChat.tsx
Comment (turt2live) [reply src/components/structures/MatrixChat.tsx:1223]: we need a comment to refer to that comment then, please :)

**Review by turt2live — COMMENTED** [review]
aside from the missing comment and possible product signoff, lgtm

**Comment by daniellekirkwood** [comment]
> > > Generally this lgtm, but it doesn't seem to be informing anything yet? Should this be behind a labs flag while we build out the stuff that uses the selection for something?
> > 
> > 
> > Our mobile apps have already implemented this screen. The primary benefit of this for now is that we immediately get the posthog metrics of which group a user belongs to.
> > Actual user-facing changes will happen with the new user onboarding checklist later on.
> 
> has Product signed off on this? I'm a bit concerned it affects the user experience in a net-negative way to not have follow up questions at the moment.

Yep! 👍  Thanks, @turt2live. We want to better understand the actions of users coming in to Element and will use that knowledge to ensure that when we personalise product based on the users answer to this question we get the personalisation right. It should be temporary and not a detrimental-to-experience question :)

**Review by turt2live — APPROVED** [review]
:shipit:

**Review by niquewoodhouse — APPROVED** [review]
Thanks, I think https://github.com/vector-im/element-web/issues/22817 is something we should address before this merges because its part of that flue

**Comment by daniellekirkwood** [comment]
> Thanks, this looks good. I think [vector-im/element-web#22817](https://github.com/vector-im/element-web/issues/22817) is something we should address before this merges because its part of that flue

I think it's something we should address but I don't see it as a blocker. We knew that the different background could cause some issues/confusion. Let's launch and take 22817 on as a bug.

**Pull request merged**
